cdk/lib/constructs/pollerLambda.ts

```diff
@@ -0,0 +1,176 @@
+import type { GuStack } from '@guardian/cdk/lib/constructs/core';
+import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
+import { aws_sqs, Duration } from 'aws-cdk-lib';
+import { RecursiveLoop } from 'aws-cdk-lib/aws-lambda';
+import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
+import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
+import type { PollerConfig, PollerId } from '../../../shared/pollers';
+import {
+	getPollerSecretName,
+	POLLER_LAMBDA_ENV_VAR_KEYS,
+	pollerIdToLambdaAppName,
+} from '../../../shared/pollers';
+import { LAMBDA_ARCHITECTURE, LAMBDA_RUNTIME } from '../constants';
+import { GuAlarm } from '@guardian/cdk/lib/constructs/cloudwatch';
+import {
+	ComparisonOperator,
+	Metric,
+	TreatMissingData,
+} from 'aws-cdk-lib/aws-cloudwatch';
+
+interface PollerLambdaProps {
+	pollerId: PollerId;
+	pollerConfig: PollerConfig;
+	ingestionLambdaQueue: aws_sqs.Queue;
+	alarmSnsTopicName: string;
+}
+
+export class PollerLambda {
+	constructor(
+		scope: GuStack,
+		{
+			pollerId,
+			ingestionLambdaQueue,
+			pollerConfig,
+			alarmSnsTopicName,
+		}: PollerLambdaProps,
+	) {
+		const lambdaAppName = pollerIdToLambdaAppName(pollerId);
+
+		const secret = new Secret(scope, `${pollerId}Secret`, {
+			secretName: getPollerSecretName(scope.stage, pollerId),
+			description: `Secret for the ${pollerId} poller lambda`,
+		});
+
+		const timeout = Duration.seconds(
+			pollerConfig.overrideLambdaTimeoutSeconds ?? 60,
+		);
+
+		// we use queue here to allow lambda to call itself, but sometimes with a delay
+		const lambdaQueue = new aws_sqs.Queue(scope, `${pollerId}LambdaQueue`, {
+			queueName: `${scope.stack}-${scope.stage}-${lambdaAppName}_queue`,
+			visibilityTimeout: timeout, // must be at least the same as the lambda
+			retentionPeriod: Duration.minutes(5),
+			// TODO dead letter queue
+			// TODO consider setting retry count to zero
+		});
+
+		if (
+			pollerConfig.idealFrequencyInSeconds &&
+			pollerConfig.idealFrequencyInSeconds > Duration.minutes(15).toSeconds()
+		) {
+			throw Error(
+				"SQS' max delay is 15mins, so if you really mean to poll so infrequently (e.g. hourly or daily), then you'll need to use Cloudwatch/EventBridge rules instead of SQS ",
+			);
+		}
+
+		// only one lambda should be happening at once, however the lambda queues up its next execution whilst it's still running
+		// so concurrency of 2 should prevent throttling, but also guard against it going haywire
+		const reservedConcurrentExecutions = 2;
```

Comment:
I don't follow what's wrong with setting reserved concurrency to 1 - yes the throttling metric will be non-zero but if that's expected behaviour then we can simply ignore the metric, and use a more useful metric to detect poor performance (ie. queue depth or age of oldest queue item, etc.)

Reply:
it's a fair question. I think the queue metrics are a bit laggy and the also the GuLambda construct gives us an easy alarm for throttling. I don't love the idea that it would always be throttling, as I think it might attempt retries on throttle. on balance I think this limit of two is the cleanest